doc/SONiC_YANG_Model_Guidelines.md

```diff
@@ -6,46 +6,153 @@
 
-Example: Table INTERFACE will translate to container INTERFACE.
+### 1. All schema definitions related to a feature should be written in a single YANG model file. YANG model file is named as 'sonic-{feature}.yang', e.g. for ACL feature sonic-acl.yang is the file name. It is mandatory to define a top level YANG container named as 'sonic-{feature}' i.e. same as YANG model name. For example, sonic-acl.yang should have 'sonic-acl' as top level container. All other definition should be written inside the top level container.
+
+Example :
+####  YANG
+```
+module sonic-acl {
+	container sonic-acl {
+		.....
+		.....
+	}
+}
+```
+
+
+### 2  It is best to categorize YANG Modules based on a networking components. For example, it is good to have separate modules for VLAN,  ACL, PORT and IP-ADDRESSES etc.
+```
+sonic-acl.yang
+sonic-interface.yang
+sonic-port.yang
+sonic-vlan.yang
+```
+
+### 3. Define namespace as "http://github.com/Azure/{model-name}" and define a shorter(preferably within 3-4 letters) namespace prefix. Use the prefix in XPath for referring to any node present in other model.
+
+Example :
+####  YANG
+```
+module sonic-acl {
+	namespace "http://github.com/Azure/sonic-acl";
+	prefix sacl;
+	.....
+	.....
+}
+```
+
+### 4. Use 'revision' to record revision history whenever YANG model is changed. Updating revision with appropriate details helps tracking the changes in the model.
+
+Example :
+
+#### YANG
+```
+module sonic-acl {
+	revision 2019-09-02 {
+		description
+			"Added must expression for ACL_RULE_LIST."; // Updated with new change details
+	}
+	
+	revision 2019-09-01 {
+		description
+			"Initial revision.";
+	}
+
+	container sonic-acl {
+		.....
+		.....
+	}
+}
+```
+
+### 5. Each primary section of ABNF.json (i.e a dictionary in ABNF.json) for example, VLAN, VLAN_MEMBER, INTERFACE  in ABNF.json will be mapped to a container in YANG model.
+
+Example: Table VLAN will translate to container VLAN.
 
 #### ABNF
 
 ```
-"INTERFACE": {
-        "Ethernet112|2a04:fxxx:40:a709::xxx/126": {
-            "scope": "global",
-            "family": "IPv6"
+"VLAN": {
+        "Vlan100": {
+             "vlanid": "107"
         }
  }
 ```
 will translate to:
 
 ####  YANG
 --
+```
+container VLAN {  //"VLAN" mapped to a container
+ list VLAN_LIST {
+   key name;
+   leaf vlanid {
+      type uint16;
+   }
+ }
+}
+```
+
+### 6. By default table is defined in CONFIG_DB, if needed use extension 'sonic-ext:db-name' for defining the table present in other Redis DB.
+
+####  YANG
+
 ```
 container INTERFACE {
+sonic-ext:db-name "APPL_DB"; //table defined in APPL_DB 
 .....
 .....
 }
 ```
 
-### 2. Each leaf in YANG module should have same name as corresponding key-fields in ABNF.json.
+### 7. The default separator used in table key  is "|". If it is different, use 'sonic-ext:key-delim {separator};' YANG extension.
```

Comment:
The config-mgmt deals with CONFIG-DB only. So let us not even have an option.

Reply:
What about state data ? I believe for showing application protocol state etc. APPL_DB would be required where I see NEIGH_TABLE, COPP_TABLE, INTF_TABLE etc. use ":" as separator instead of "|".

Reply:
State data --
We need some clarity here. 
1) This is a read-only data, filled by DB access layer.
2) No other module handles it as there is no validation/translation required.

This implies that all we need for each read-only field in YANG is, meta-data to specify the source for this data, which would include table + key. The DB access layer would know the separator to use, depending on table.

Let us have some samples and then go from there, to get the guidelines.
For now, let us not add any.

Reply:
Do you want to remove it from guideline ? 
DB access layer should not hard code separator or source of DB, it should derive from YANG annotation.
Can you please explain "Let us have some samples and then go from there, to get the guidelines." ?
Where we need to add the samples ?

Reply:
I have merged this with the step where read-only data is defined.